lib/utils/load-config.js

```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+const path = require('path');
+const { execSync } = require('node:child_process');
+
+module.exports = async (configPath) => {
+  try {
+    return require(configPath);
+  } catch (error) {
+    // Fallback to import() if the runtime supports native ESM
+    if (error.code === 'ERR_REQUIRE_ESM') {
+      try {
+        return (await import(`file:///${configPath}`)).default;
+      } catch (errorESM) {
+        // TypeScript .ts files can only be imported by Node if ts-node was registered on process startup
+        // either with CLI flags '--loader ts-node/esm --experimental-modules --es-module-specifier-resolution=node'
+        // or as environment variable 'NODE_OPTIONS="--loader ts-node/esm --es-module-specifier-resolution=node"'
+        if (errorESM.code === 'ERR_UNKNOWN_FILE_EXTENSION' && path.extname(configPath) === '.ts') {
+          // load-config-esm is an ES module that loads the config file and prints it to stdout
+          const loadConfigEsm = path.join(__dirname, 'load-config-esm');
+
+          // Execute the load-config-esm module with the config path as argument
+          const config = execSync(
```

Comment:
I think this logic should not be here, as in an earlier point (https://github.com/serverless/serverless/blob/37b29b029579191e82348afbf855590c2cb37062/lib/configuration/read.js#L111-L134) we guarantee `ts-node` is loaded.

So here, I think we should not involve any TS specific logic, but should focus on pure native ESM support

Reply:
If I understood correctly, `ts-node` won't be recognized properly in such case as it requires a runtime flag or env var (cannot be registered programmatically in this case) and that's why @zirkelc suggests such workaround here. Please correct me if I'm wrong

Reply:
I understand we address here`.ts` config case.

Another question came to me. So is this possible to write `.ts` file in way that it resolves as ESM? Isn't it always resolved to CJS in Node.js?

Reply:
Nonetheless, nesting Node.js processes like is definitely not a thing we should do. We also should not support _experimental modules_ but only _stable_

Reply:
I removed the `--experimental-modules` CLI flag as it is no longer necessary to specify it since Node v12.20 & v14: https://nodejs.org/api/esm.html#modules-ecmascript-modules

I understand that nesting Node.js processes is not an optimal solution, but this is the only way to specific the `--loader` flag for ESM: https://github.com/TypeStrong/ts-node/issues/1791

Reply:
As I check node.js documentation, there's no `--loader` option listed: https://nodejs.org/api/cli.html.

Is this experimental (not stable) option ?

Reply:
> Is there no other way to resolve the config this way? I'm honestly not a fan of resolving configuration by parsing output from another process and would treat it only as a last resort really.
> 
> @medikoo What is your opinion on this approach?

@pgrzesik I changed the implementation to use `fork()` instead of `execSync()` to start a Node.js child process. This automatically establishes an IPC channel between parent and child to communicate with each other and allows to send the Serverless config as message back to `load-config.js`. This I could avoid to print it to stdio.

https://nodejs.org/docs/latest-v16.x/api/child_process.html#child_processforkmodulepath-args-options

Reply:
> As I check node.js documentation, there's no `--loader` option listed: https://nodejs.org/api/cli.html.
> 
> Is this experimental (not stable) option ?

I couldn't find it in the docs either, but `node --help` does lists this option:

>   --loader, --experimental-loader=... : use the specified module as a custom loader

But it's alternative option name is `--experimental-loader`, so yes it is still experimental: https://nodejs.org/api/esm.html#loaders

Reply:
Given the fact that it's an experimental feature and it uses a bit controversial method of invoking a separate process just to load the TS config, I don't think we should support this approach at this point, what do you think @medikoo ?